### ISSUE TITLE
Sync fails on error 404

### ISSUE THREAD

**Issue by jdeluyck** [andrewramsay/ical_to_gcal_sync#1]
Hi,

I've been looking for a way to manually sync an ICS file into Google, as their own sync is broken beyond belief, and they still don't want to fix it.

I registered the application, and then created the json file and config. Start the script, it starts up but bombs out with:
```
googleapiclient.errors.HttpError: <HttpError 404 when requesting https://www.googleapis.com/calendar/v3/calendars/kcore.org_xxxxxxxxxxxxxxxxxxxxxxxxxx%40group.calendar.google.com/events/0c234de16509d920c12577c3003d751blotus_notes_generated?alt=json returned "Not Found">
```

The calendar in question is on google's GSuite platform, but that shouldn't really matter I think. 
I can see that it has synced some items already, but then stops.

**Comment by andrewramsay** [comment]
I don't remember running into that error before. Do you have the stack trace to show where the exception gets thrown? 

If it's managed to sync some events before hitting the problem it presumably isn't having any issues accessing the calendar itself, so I guess there might be something about a specific event that's tripping it up. No idea what it could be offhand, but if you can check which API method is being called when it happens that might give us some indication of what is going wrong.

**Comment by jdeluyck** [comment]
Unfortunately, it's not really obvious why it's failing:

```
Traceback (most recent call last):
  File "ical_to_gcal_sync.py", line 197, in <module>
    service.events().update(calendarId=CALENDAR_ID, eventId=gcal_event['id'], body=gcal_event).execute()
  File "/usr/lib/python2.7/site-packages/oauth2client/_helpers.py", line 133, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/googleapiclient/http.py", line 840, in execute
    raise HttpError(resp, content, uri=self.uri)
googleapiclient.errors.HttpError: <HttpError 404 when requesting https://www.googleapis.com/calendar/v3/calendars/kcore.org_xxxxxxxxxxx%40group.calendar.google.com/events/0c234de16509d920c12577c3003d751blotus_notes_generated?alt=json returned "Not Found">
```

Which hits except: block, so it's not really clear.

```
except:
                time.sleep(API_SLEEP_TIME)
                service.events().update(calendarId=CALENDAR_ID, eventId=gcal_event['id'], body=gcal_event).execute()
```

**Comment by jdeluyck** [comment]
I've added some more details to the try/catch block, and now I'm getting an 
```
Traceback (most recent call last):
  File "ical_to_gcal_sync.py", line 195, in <module>
    service.events().insert(calendarId=CALENDAR_ID, body=gcal_event).execute()
  File "/usr/lib/python2.7/site-packages/oauth2client/_helpers.py", line 133, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/googleapiclient/http.py", line 840, in execute
    raise HttpError(resp, content, uri=self.uri)
googleapiclient.errors.HttpError: <HttpError 400 when requesting https://www.googleapis.com/calendar/v3/calendars/kcore.org_xxxxxxxxxxxxxxxxxxxxxxx%40group.calendar.google.com/events?alt=json returned "Invalid resource id value.">
```
After digging around in the ICS file, it's this event that causes the error:
```
< RRULE:FREQ=YEARLY;UNTIL=20181110T230000Z;INTERVAL=1;BYMONTHDAY=11;BYMONTH=1
<  1
< UID:0C234DE16509D920C12577C3003D751B-Lotus_Notes_Generated
< SUMMARY:Wapenstilstand
< DTSTART;VALUE=DATE:20181111
< DTEND;VALUE=DATE:20181112
< CLASS:PUBLIC
< PRIORITY:5
< DTSTAMP:20180802T171618Z
< TRANSP:OPAQUE
< STATUS:CONFIRMED
< X-MICROSOFT-CDO-BUSYSTATUS:BUSY
< X-MICROSOFT-CDO-INTENDEDSTATUS:BUSY
< X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
< X-MICROSOFT-CDO-IMPORTANCE:1
< X-MICROSOFT-CDO-INSTTYPE:1
< X-MICROSOFT-DONOTFORWARDMEETING:FALSE
< X-MICROSOFT-DISALLOW-COUNTER:FALSE
< END:VEVENT
```
I'm not sure what the '1' is doing on a single row, but kicking out this event let the import complete.

https://github.com/jdeluyck/ical_to_gcal_sync is my fork.

**Comment by andrewramsay** [comment]
As I said in the readme I really didn't bother testing this beyond "it works for me", so I'm not surprised there are bugs like this to find :)

After checking the the details of the API, I think the "-Lotus_Notes_Generated" text appended to the UID could be the problem - does that appear on any other events in your .ics file? I did a quick test with my own .ics data, and if I edit one of the UID values and append that text, it throws up the same exception you are seeing. 

The [API docs for event UIDs](https://developers.google.com/calendar/v3/reference/events) states "characters allowed in the ID are those used in base32hex encoding, i.e. lowercase letters a-v and digits 0-9". My script converts to lowercase at least, but the iCal source I'm using always generates standard UUID-formatted IDs, so I didn't add any more validation in there. 

If you can modify/remove that appended text somehow I think it should fix the problem, the rules on what constitutes a valid GCal UID are quite relaxed otherwise. Maybe just extend convert_uid to make sure there are no invalid characters left?

**Comment by jdeluyck** [comment]
Sounds like a plan. Are you interested in pull requests?

The only other that is not an UUID is `UID:368941221@logmein.com`.

**Comment by andrewramsay** [comment]
Sure, feel free to submit one!

**Comment by andrewramsay** [comment]
OK, this should be fixed by #2 now

Thanks for that, am sure it'll help anyone else who decides to try using this!
